File: 2025/golang/day3/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func solve(input string) {
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	total := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		lenLine := len(line)
		maxIdx := 0
		secMaxIdx := 1
		if line[1] > line[0] {
			maxIdx = 1
			secMaxIdx = 0
		}
		for i := 2; i < lenLine-1; i++ {
			if line[i] > line[maxIdx] {
				secMaxIdx = lenLine - 1
				maxIdx = i
			} else if line[i] > line[secMaxIdx] {
				secMaxIdx = i
			}
		}

		if line[secMaxIdx] < line[lenLine-1] {
			secMaxIdx = lenLine - 1
		}

		num, err := strconv.Atoi(string(line[maxIdx]) + string(line[secMaxIdx]))
		if err != nil {
			panic(err)
		}
		total += num

	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	fmt.Println("Total: ", total)

}

func main() {
	// input := "input_test.txt" // 357
	input := "input.txt" // 17085
	solve(input)
}

File: 2025/python/day3/part1.py
def main(input: str) -> None:
    total = 0
    with open(input, "r") as f:
        for line in f.readlines():
            line = line.strip()
            if int(line[0]) > int(line[1]):
                max_idx, sec_max_idx = 0, 1
            else:
                max_idx, sec_max_idx = 1, 0

            for idx, digit in enumerate(line[2:-1]):  # We suppose that len(line) > 2
                if int(digit) > int(line[max_idx]):
                    sec_max_idx = len(line) - 1
                    max_idx = idx
                elif int(digit) > int(line[sec_max_idx]):
                    sec_max_idx = idx

            if int(line[-1]) > int(line[sec_max_idx]):
                sec_max_idx = len(line) - 1

            total += int(f"{line[max_idx]}{line[sec_max_idx]}")
            print(line)
            print(max_idx)
            print(sec_max_idx)
            print(int(f"{line[max_idx]}{line[sec_max_idx]}"))

    print("total:", total)


if __name__ == "__main__":
    input_str = "input_test.txt"
    # input_str = "input.txt"
    main(input_str)
